resources/testharness.js

```diff
@@ -1621,29 +1632,99 @@ policies and contribution forms [3].
      */
     Test.prototype.cleanup = function() {
         var error_count = 0;
-        var total;
+        var bad_value_count = 0;
+        function on_error() {
+            error_count += 1;
+            // Set test phase immediately so that tests declared within
+            // subsequent cleanup functions are not run.
+            tests.phase = tests.phases.ABORTED;
+        }
+        var this_obj = this;
+        var results = [];
+
+        this.phase = this.phases.CLEANING;
 
         forEach(this.cleanup_callbacks,
                 function(cleanup_callback) {
+                    var result;
+
                     try {
-                        cleanup_callback();
+                        result = cleanup_callback();
                     } catch (e) {
-                        // Set test phase immediately so that tests declared
-                        // within subsequent cleanup functions are not run.
-                        tests.phase = tests.phases.ABORTED;
-                        error_count += 1;
+                        on_error();
+                        return;
                     }
+
+                    results.push(result);
                 });
 
-        if (error_count > 0) {
-            total = this._user_defined_cleanup_count;
+        if (!this._is_promise_test) {
+            forEach(results,
+                    function(result) {
+                        if (result !== undefined) {
+                            bad_value_count += 1;
+                        }
+                    });
+            cleanup_done(this_obj, error_count, bad_value_count);
+        } else {
+            all_async(results,
+                      function(result, done) {
+                          if (result && typeof result.then === "function") {
+                              result
+                                  .then(null, on_error)
+                                  .then(done);
+                          } else {
+                              if (result !== undefined) {
+                                  bad_value_count += 1;
+                              }
+
+                              done();
+                          }
+                      },
+                      function() {
+                          cleanup_done(this_obj, error_count, bad_value_count);
+                      });
+        }
+    };
+
+    function cleanup_done(test, error_count, bad_value_count) {
+        if (error_count || bad_value_count) {
+            var total = test._user_defined_cleanup_count;
+
+            tests.phase = tests.phases.ABORTED;
+            forEach(tests.tests,
+                    function(test) {
+                        test.phase = test.phases.COMPLETE
```

Comment:
This wasn't done before, why do the individual tests need to be set to `COMPLETE` here?

Reply:
The short answer is: so the tests pass.

The long answer is: to enable a implicit and brittle solution to function. It
turns out that the original patch set the properties and relied on the
following code to react to the new state:

```js
test_environment.add_on_loaded_callback(function() {
    if (this_obj.all_done()) {
        this_obj.complete();
    }
});
```

So as written, aborting was not conducted imperatively. The actual
implementation could be described in plain English as, "set the properties and
wait for a DOM event to notice that change." This works in the highly
controlled "`testharness.js` test" environment, but it may not apply generally.

I've updated the implementation and added a new test to verify. (This has
obviated the code you identified here.)

Reply:
This attention to detail is much appreciated, it's easy to write brittle code, but hard to understand the resulting failures and even harder to make the code less brittle over time. Yay!